UI tests for Private browsing settings sub-menu

sv-ohorvath opened the pull request:
### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

sv-ohorvath commented:
Currently blocked by: 
https://github.com/mozilla-mobile/fenix/issues/8354

AaronMT commented:
@sv-ohorvath Is this ready for review now?

sv-ohorvath commented:
> @sv-ohorvath Is this ready for review now?
@AaronMT 
Yes, I think it is,  I was waiting to run it a few times before asking a review. It passed the last 2 days, so please have a look, especially if I should add anything else to prevent them from flakiness.

AaronMT commented on app/src/androidTest/java/org/mozilla/fenix/ui/SettingsPrivacyTest.kt:
```diff
@@ -138,21 +143,21 @@ class SettingsPrivacyTest {
+        }.openThreeDotMenu {
+        }.openSettings {
+            TestHelper.scrollToElementByText("Logins and passwords")
+        }.openLoginsAndPasswordSubMenu {
```
I would add the scroll line into the open function

AaronMT commented on app/src/androidTest/java/org/mozilla/fenix/ui/SettingsPrivacyTest.kt:
```diff
@@ -138,21 +143,21 @@ class SettingsPrivacyTest {
+        }.openSettings {
+            TestHelper.scrollToElementByText("Logins and passwords")
+        }.openLoginsAndPasswordSubMenu {
+            verifyDefaultView()
```
Should have a more descriptive name I think

AaronMT commented on app/src/androidTest/java/org/mozilla/fenix/ui/SettingsPrivacyTest.kt:
```diff
@@ -194,6 +199,98 @@ class SettingsPrivacyTest {
+    fun openExternalLinksInPrivateTest() {
+        setOpenLinksInPrivateOn()
+
+        openAppFromExternalLink("https://google.com")
```
Is it possible to use the MockwebServer and a generic asset page instead?

AaronMT commented on app/src/androidTest/java/org/mozilla/fenix/ui/SettingsPrivacyTest.kt:
```diff
@@ -194,6 +199,98 @@ class SettingsPrivacyTest {
+
+        setOpenLinksInPrivateOff()
+
+        restartApp(activityTestRule)
```
Does Espresso need to wait here? What happens if this takes a while? ('m not familiar with `finishActivity` and how it works in the context of the instrumentation runner)

sv-ohorvath replied:
Do you think a waitForIdle would be enough?

AaronMT replied:
I missed this. I'm not too familiar with either, I think it might be worth experimenting with each.

AaronMT commented on app/src/androidTest/java/org/mozilla/fenix/ui/SettingsPrivacyTest.kt:
```diff
@@ -365,3 +462,27 @@ class SettingsPrivacyTest {
 }
+
+private fun setOpenLinksInPrivateOn() {
+
```
nit: extra line (remember to run `ktlint` in Gradle)

AaronMT commented on app/src/androidTest/java/org/mozilla/fenix/ui/SettingsPrivacyTest.kt:
```diff
@@ -365,3 +462,27 @@ class SettingsPrivacyTest {
+    }.goBack {
+    }.goBack {
+    }
+}
```
Should probably verify that we're back on the home screen here

AaronMT commented on app/src/androidTest/java/org/mozilla/fenix/ui/SettingsPrivacyTest.kt:
```diff
@@ -365,3 +462,27 @@ class SettingsPrivacyTest {
+        clickOpenLinksInPrivateTabSwitch()
+    }.goBack {
+    }.goBack {
+    }
```
nit: goBack {}

AaronMT commented on app/src/androidTest/java/org/mozilla/fenix/ui/SettingsPrivacyTest.kt:
```diff
@@ -365,3 +462,27 @@ class SettingsPrivacyTest {
+        clickOpenLinksInPrivateTabSwitch()
+        verifyOpenLinksInPrivateTabOff()
+    }.goBack {
+    }.goBack {}
```
Same here

AaronMT commented on app/src/androidTest/java/org/mozilla/fenix/ui/robots/AddToHomeScreenRobot.kt:
```diff
@@ -0,0 +1,86 @@
+        }.enterURLAndEnterToBrowser(url) {
+        }.openThreeDotMenu {
+        }.openAddToHomeScreen {
+            mDevice.waitNotNull(Until.findObject(By.text("Add to Home screen")), waitingTime)
```
I'm confused why you're mixing in robot usage here outside of a test? Shouldn't this function just confirm the existence of the popup dialog, the shortcut title and perform the click? The test should handle getting there through the menus.

AaronMT commented on app/src/androidTest/java/org/mozilla/fenix/ui/robots/AddToHomeScreenRobot.kt:
```diff
@@ -0,0 +1,86 @@
+                Until.findObject(By.text(title)),
+                waitingTime
+            )
+            mDevice.findObject((UiSelector().text(title))).clickAndWaitForNewWindow()
```
Does `clickAndWaitForNewWindow` need a provided timeout value?

sv-ohorvath replied:
"Waits for window transitions that would typically take longer than the usual default timeouts" - I don't see anywhere what is the amount of time it waits...
I haven't seen any issues yet without a specified time either, but I could add the TestAssetHelper.waitingTime - would that be better?

AaronMT replied:
Yeah I would use that

AaronMT commented on app/src/androidTest/java/org/mozilla/fenix/ui/robots/SettingsSubMenuPrivateBrowsingRobot.kt:
```diff
@@ -0,0 +1,108 @@
+    fun addPrivateShortcutToHomescreen() {
+        addPrivateBrowsingShortcutButton().click()
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            addAutomaticallyButton().click()
```
Does this work for API 25 through API 28?

sv-ohorvath commented:
> There's a lot going on in these tests. I'm concerned for how these work on AVDs and physical devices as there are API differences, and going to the home screen and back. I would rebase and get a run on Firebase and see how it performs a couple times.
@AaronMT I've addressed your comments and made some small changes. Please have another look if you can. I'll leave this open another 1-2 days and re-run the tests a few times then see what happens. Thank you!

AaronMT commented:
> > There's a lot going on in these tests. I'm concerned for how these work on AVDs and physical devices as there are API differences, and going to the home screen and back. I would rebase and get a run on Firebase and see how it performs a couple times.
> 
> @AaronMT I've addressed your comments and made some small changes. Please have another look if you can. I'll leave this open another 1-2 days and re-run the tests a few times then see what happens. Thank you!

Sound's good. I'd rebase as master was broken from an AC change.

AaronMT commented:
Let's see how it runs after another rebase once the build is green again.

sv-ohorvath commented:
PR approved, ran the checks about 10 times and the tests didn't fail so I'm going to merge this.
